### PR TITLE
[G2M] VISX/Plotly/D3 exp

### PR DESCRIPTION
**Pull request by DoParkEQ**
- modularized cluster, gauge bar and gauge arc from overlord
- reorg/ad sample data 
- added visx/plotly version of basic charts

### CODE REVIEW

File: src/components/d3/gauge-bar.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+  const percentage = data[dataKey.x2] ? data[dataKey.x1] / data[dataKey.x2] * 100 : 0
+
+  const axisRange = axis.dynamicRange ? [axis.range[0], Math.ceil(percentage + 1.0)] : [axis.range[0], axis.range[1]]
+
+  const xScale = d3.scaleLinear()
+    .domain(axisRange)
+    .range([0, width])
+
+  console.log(percentage)
```
Comment (geoerika) [inline src/components/d3/gauge-bar.js:22]: remove console

File: src/components/visx/Pie.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+const renderPie = (data, config, pie) => {
+
+  //TODO: add text auto color contrast
+  const { dataKey: { item }, colorRange } = config
+  const pieColor = scaleOrdinal({
+    domain: data.map((d) => d.item),
+    range: colorRange,
+  })
+  console.log(pie)
```
Comment (geoerika) [inline src/components/visx/Pie.js:18]: remove console

File: src/components/d3/gauge-arc.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+          return //deactivate tweening for the remainder value
+        }
+        var i = d3.interpolate(d.startAngle + 0.1, d.endAngle)
+        return function (t) {
+
+          d.endAngle = i(t)
+          return arcPath(d)
+        }
+      })
+  }, [data])
```
Comment (geoerika) [inline src/components/d3/gauge-arc.js:33]: dependency lists seem to miss variables, can you check them all please?

**Review by geoerika — CHANGES_REQUESTED** [review]
Can we clean the data files so we don't have them all with lint warnings please? There are lots of commas missing and doublequotes.

File: .depcheckrc [hunk 4]
```diff
@@ -8,5 +8,7 @@ ignores: [
     "@babel/*",
     "@storybook/*",
     "size-limit",
-    "@size-limit/*"
+    "@size-limit/*",
+    "plotly.js",
```
Comment (geoerika) [inline .depcheckrc:12]: Question; why did you need to add plotly to this list? Just wondering...

File: stories/cluster.stories.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    color: args.color,
+    currentGroup: args.currentGroup,
+    dataKey: {
+      node: 'GeoCohortListID',
+      radius: args.valueKey,
+    },
+    mode: {
+      showCommonNodes: args.commonNodes,
+      groups: [1, 2],
+      colors: ['#a9ff91', '#4278ff', '#dc91ff']
```
Comment (geoerika) [inline stories/cluster.stories.js:74]: missing comma here

File: stories/cluster.stories.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+      style: {
+        position: 'absolute',
+        minWidth: 80,
+        backgroundColor: '#fff',
+        padding: 16,
+        border: 'none',
+        borderRadius: 4,
+        boxShadow: '0 2px 8px 0 rgba(12, 12, 13, 0.15)',
+        visibility: 'hidden',
+      }
```
Comment (geoerika) [inline stories/cluster.stories.js:88]: add comma here too

File: stories/cluster.stories.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    valueKey: 'Imps',
+    remainderKey: 'Bids',
+    commonNodes: false,
+  },
+  argTypes: {
+    width: {
+      control: {
+        type: 'number',
+        options: [0, 500],
+      }
```
Comment (geoerika) [inline stories/cluster.stories.js:23]: check all missing trailing commas in this big object...

File: stories/gauge-arc.stories.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    remainderKey: 'Bids',
+    animation: true,
+    duration: 1000,
+  },
+  argTypes: {
+    width: {
+      control: {
+        type: 'number',
+        options: [0, 500],
+      }
```
Comment (geoerika) [inline stories/gauge-arc.stories.js:24]: I think we need trailing commas in lots of places in this object as well.

**Review by geoerika — CHANGES_REQUESTED** [review]

File: src/components/d3/gauge-arc.js [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+            var i = d3.interpolate(d.startAngle + 0.1, d.endAngle)
+            return function (t) {
+
+              d.endAngle = i(t)
+              return arcPath(d)
+            }
+          })
+      }
+    }
+  }, [svgRef.current, arcsData])
```
[the remark this answers is not in the document]
Comment (DoParkEQ) [reply src/components/d3/gauge-arc.js:56]: I don't see any usecases which the transition prop changes after the first cycle as the animation should be triggered only once.

File: .depcheckrc (hunk 4 above)
Comment (DoParkEQ) [reply .depcheckrc:12]: depcheck raises the flag for plotly with an unknown reason though I didn't see any issues. I am assuming it is due to its peer dependency which is react-plotly.

**Review by geoerika — CHANGES_REQUESTED** [review]

File: src/components/d3/gauge-arc.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (DoParkEQ) [reply src/components/d3/gauge-arc.js:56]: My understanding is that the rules are only the recommendation made without a context of the code itself which can alter the way it functions, therefore it is only set to give us warning not error. For the same component, adding all dependencies from what eslint recommended messed up with the rendering cycle so I had to revert in overlord.

[the remark this answers is not in the document]
Comment (geoerika) [reply src/components/d3/gauge-arc.js:56]: Well, it is not what react docs actually state. This is from the official React page, at: https://reactjs.org/docs/hooks-effect.html
`Note
If you use this optimization, make sure the array includes all values from the component scope (such as props and state) that change over time and that are used by the effect. Otherwise, your code will reference stale values from previous renders. Learn more about how to deal with functions and what to do when the array changes too often.

If you want to run an effect and clean it up only once (on mount and unmount), you can pass an empty array ([]) as a second argument. This tells React that your effect doesn’t depend on any values from props or state, so it never needs to re-run. This isn’t handled as a special case — it follows directly from how the dependencies array always works.

If you pass an empty array ([]), the props and state inside the effect will always have their initial values. While passing [] as the second argument is closer to the familiar componentDidMount and componentWillUnmount mental model, there are usually better solutions to avoid re-running effects too often. Also, don’t forget that React defers running useEffect until after the browser has painted, so doing extra work is less of a problem.

We recommend using the exhaustive-deps rule as part of our eslint-plugin-react-hooks package. It warns when dependencies are specified incorrectly and suggests a fix.
`
Also, I have faced this in many code reviews in the past year and always found a better solution. I don't know if it is possible in this case. Maybe @woozyking can give a better input.

File: src/components/d3/gauge-arc.js [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+  backgroundColor: '#cdcdcd',
+  transition: {
+    animation: false,
+    duration: 0,
+  },
+}
+
+
+export default React.memo(GaugeArc)
+
```
[the remark this answers is not in the document]
Comment (geoerika) [reply src/components/d3/gauge-arc.js:105]: This hasn't been fixed

**Review by geoerika — CHANGES_REQUESTED** [review]

File: src/components/d3/gauge-arc.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (DoParkEQ) [reply src/components/d3/gauge-arc.js:105]: It has a single space instead of double. Are you still seeing that?

File: src/components/d3/gauge-arc.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (DoParkEQ) [reply src/components/d3/gauge-arc.js:56]: It states that the purpose of including all values used in useEffect is to prevent itself from not re-rendering when the value updates. Now when you take a look at this deeper, many of our system cause an unnecessary rendering due to its data flow. Geocohort report tab for example, it re renders when it fetches other non-geocohort data such as conversion or overview, and react recognizes it as a mutated value when geocohort data itself hasn't changed. So in order to manually orchestrate the rendering cycle in harmony, some manual adjustment has to be imposed. Selective deps is the one way. Another way of addressing that while avoiding warning is to add conditional statement but I am not a big supporter of adding bunch of them as it defeats the purpose of hooks dependencies.

[the remark this answers is not in the document]
Comment (DoParkEQ) [reply src/components/d3/gauge-arc.js:56]: It states that the reason that they recommend adding all values is to prevent the case where component does not update due to its stale value. However in this case, I purposefully do not want to update component because it has to animate only once in the first rendering cycle. One way to address I know would be imposing conditional statement inside of useEffect but I personally think that having too many of them are defeating the purpose of having deps list while it makes the code bulky. Therefore I am leaning towards this approach.

File: src/components/d3/gauge-arc.js [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+  const isAscending = value > remainder
+
+  const pieSize = width > height ? height / 2 : width / 2
+
+  const lineThickness = pieSize * 0.2
+
+  const arcsGenerator = d3.pie().sortValues(function (a, b) { return isAscending ? b - a : a - b })
+
+  const arcsData = useMemo(() => arcsGenerator([value, remainder]), [data])
```
Comment (geoerika) [inline src/components/d3/gauge-arc.js:23]: I think here you are safe replacing data with'arcsGenerator', 'remainder', and 'value'

**Review by geoerika — CHANGES_REQUESTED** [review]

File: src/components/d3/cluster.js [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+
+const radiusMin = 5
+const radiusMax = 15
+const start = 0
+
+const Cluster = ({ width, height, data, config }) => {
+
+  const { dataKey, color, currentGroup, tooltip, clusterMaxLength, mode } = config
+  const end = clusterMaxLength > 0 ? clusterMaxLength : data.length
+  const clusters = useMemo(() => data.sort((a, b) => d3.descending(a[dataKey.radius], b[dataKey.radius])).slice(start, end), [data, dataKey.radius])
```
Comment (geoerika) [inline src/components/d3/cluster.js:14]: Is this one also giving troubles if you add 'end' to it?

**Review by geoerika — CHANGES_REQUESTED** [review]

**Review by geoerika — APPROVED** [review]
